src/openms/include/OpenMS/KERNEL/MSExperiment.h

```diff
@@ -336,62 +337,180 @@ namespace OpenMS
         {
           t = (float)it.getRT();
           rt.push_back(t);
+          mz.push_back(std::vector<float>());
+          intensity.push_back(std::vector<float>());
         }
         mz.back().push_back((float)it->getMZ());
         intensity.back().push_back(it->getIntensity());
       }
     }
 
-    // for fast pyOpenMS access to MS1 peak data in format: [rt, [mz, intensity, ion mobility]]
-    void get2DPeakData(CoordinateType min_rt, CoordinateType max_rt, CoordinateType min_mz, CoordinateType max_mz, 
+    /* @brief Retrieves the peak data in the given mz-rt range and store data spectrum-wise in separate arrays.
+     * 
+     * For fast pyOpenMS access to MS1 peak data in format: [rt, [mz, intensity, ion mobility]]
+     * 
+     * @param min_rt The minimum retention time.
+     * @param max_rt The maximum retention time.
+     * @param min_mz The minimum m/z value.
+     * @param max_mz The maximum m/z value.
+     * @param ms_level The MS level of the spectra to consider.
+     * @param rt The vector to store the retention times in.
+     * @param mz The vector to store the m/z values in.
+     * @param intensity The vector to store the intensities in.
+     * @param ion_mobility The vector to store the ion mobility values in.
+    */
+    void get2DPeakDataIMPerSpectrum(
+      CoordinateType min_rt, 
+      CoordinateType max_rt, 
+      CoordinateType min_mz, 
+      CoordinateType max_mz,      
       std::vector<float>& rt, 
+      Size ms_level,
       std::vector<std::vector<float>>& mz,
       std::vector<std::vector<float>>& intensity, 
       std::vector<std::vector<float>>& ion_mobility) const
     {
+      DriftTimeUnit unit;
+      std::vector<float> im;
       float t = -1.0;
-      for (auto it = areaBeginConst(min_rt, max_rt, min_mz, max_mz); it != areaEndConst(); ++it)
+      for (auto it = areaBeginConst(min_rt, max_rt, min_mz, max_mz, ms_level); it != areaEndConst(); ++it)
       {
         if (it.getRT() != t)
         {
           t = (float)it.getRT();
           rt.push_back(t);
+          std::tie(unit, im) = it.getSpectrum().maybeGetIMData();
+          mz.push_back(std::vector<float>());
+          intensity.push_back(std::vector<float>());
+          ion_mobility.push_back(std::vector<float>());
         }
-        
-        const MSSpectrum& spectrum = it.getSpectrum();
-        bool has_IM = spectrum.containsIMData();
-        float peak_IM{-1.0f};
-        if (has_IM)
+
+        if (unit != DriftTimeUnit::NONE)
         {
-          const auto& im_data = spectrum.getIMData();
           const Size peak_index = it.getPeakIndex().peak;
-          if (spectrum.getFloatDataArrays()[im_data.first].size() == spectrum.size())
-          {
-            peak_IM = spectrum.getFloatDataArrays()[im_data.first][peak_index];
-          }          
+          ion_mobility.back().push_back(im[peak_index]);
+        }
+        else
+        {
+          ion_mobility.back().push_back(-1.0);
```

Comment:
@jpfeuffer fine to use -1 if no IM set?

Reply:
Hmm good question. I guess numpy on the python side could handle nan as well.

Reply:
@axelwalter or @jcharkow  any opinion or convention  here?

Reply:
Not sure if an error should be thrown if ion mobility not found? In my opinion the data might be corrupt if that is the case.

Reply:
Fine with me if there is never a use case where only a part of the spectra have IM values.

Reply:
I have not seen a case where ion mobility is not present, however I'm not sure if this is a case for others as I only work with timsTOF

Reply:
no strong opinion here if it is only for peak extraction I would also be fine with a placeholder.